### DISCUSSION TITLE
Guitar Effects Pedal with ESP32 and Arduino Audio tools (a.k.a ADC in to DAC out)

### DISCUSSION THREAD

**Discussion by FCJFB** [pschatzmann/arduino-audio-tools#225] [Q&A] [answered]
Hi everyone,

I wanted to build a really simple guitar pedal which uses the built in ADC and DAC from the ESP32, but trying to set up the variables of 
`AnalogAudioStream in,out;  `to the StreamCopy function returns me after booting up 

> "(399714) I2S: /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/driver/i2s.c:1214 (i2s_write):tx NULL
> [E] AnalogAudioESP32.h : 417 - size_t audio_tools::AnalogAudioStream::outputMono(const void*, size_t): 0"

 instead of the usual

>  "AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops"

 , making me start to believe that this idea is not feasible with the barebone hardware of the ESP32. Maybe its also that im coding it wrong

So my question is, is this idea feasible with my current hardware?
If not, would an external I2S DAC solve this issue as it would use a different channel/port?
How else could I implement this idea?

Here is my code in case I am missing something:
```
#include "Arduino.h"
#include "AudioTools.h"


uint16_t sample_rate=44100;

AnalogAudioStream in,out; 
StreamCopy copier(out, in);                             // copies sound into i2s

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // RX automatically uses port 0 with pin GPIO34
  auto cfgRx = in.defaultConfig(RX_MODE);
  cfgRx.sample_rate = sample_rate;
  in.begin(cfgRx);
  // start the analog output to pin 25
  auto config = out.defaultConfig(TX_MODE);
  config.sample_rate = sample_rate; 
  out.begin(config);

}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();
}
```

**Comment by pschatzmann** [comment, marked as answer]
Analog input requires port 0 as does Analog Output: so it is technically not possible to do analog input and output on an ESP32 using the I2S interface.

There is no such restriction if you do regular (digital) I2S for input or output.

**Reply by FCJFB** [reply]
So the analogADC to for example a max98357a module would be possible if im understanding correctly?

**Reply by pschatzmann** [reply]
Hmm, the max98357a is a DAC / Amplifier and not a ADC.

**Reply by FCJFB** [reply]
Sorry for not being clear, I meant using the analog ADC from the ESP32 to send this data through the code and outputting it on the max98357A module would be possible?
(basically it would be the streams-adc-i2s example to getting the tone without being modified through)

**Comment by pschatzmann** [comment]
I think it will not be great, but technically it will work. 
How do you plan to connect the Guitar to the ADC ?

**Reply by FCJFB** [reply]
Well, my thoughts were through a jack connector which leads to a biasing point where the input with a voltage divider will be elevated to 1.3V  (middlepoint of ESP32 ADC voltage) and inputting this into a unity gain amplifier just in case. Maybe some decoupling condensers here and there to stabilize a bit the power supply noise, even a zener regulator would be possible, but all this would be to worry about after knowing software wise it woudl work.
